### PR TITLE
chore(adr): add ADR for expanding subplat metadata

### PR DESCRIPTION
**Pull request by chenba**
Because:
 * we want to use more metadata than the limit Stripe placed on its
   metadata

This commit:
 * add an ADR to document our choices and decisions

### CODE REVIEW

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (chenba) [inline docs/adr/0026-expand-subscription-platform-metadata.md:4]: @bbangert @biancadanforth included you since I've chatted with you about this already.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (meandavejustice) [inline docs/adr/0026-expand-subscription-platform-metadata.md:50]: We can also add validation which should appear in the console to product managers, https://firebase.google.com/docs/rules/data-validation

This is something we are lacking with our new metadata validation

**Review by meandavejustice — APPROVED** [review]

**Comment by meandavejustice** [comment]
I wasn't a part of this discussion, but It sounds like the decision has been made. The only thing I would like in addition to the pros listed above, would be a platform that allows for data validation on input. Our current validation implementation #8920 is lacking in that it does not let the product managers know that the data is invalid directly and has us building out other scripts in order to notify them and the engineers of the failure. It looks like firestore may support something like this, I linked above..

**Comment by chenba** [comment]
>It sounds like the decision has been made.

@meandavejustice I've been confused by the ADR process in the past as well, but the decision has not made.  An ADR PR is really a request for comments.  (The recent email templates ADR is actually a great example; the chosen option was not the one in the original draft.)

**Comment by lmorchard** [comment]
> I've been confused by the ADR process in the past as well, but the decision has not made. 

Pardon the drive-by comment, but: Yeah, the ADR decision isn't really properly made until the PR gets merged. Folks can request changes & block merge in the meantime.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (bbangert) [reply docs/adr/0026-expand-subscription-platform-metadata.md:50]: Wow, I hadn't seen that feature, that's slick!

**Review by bbangert — APPROVED** [review]
LGTM.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (chenba) [reply docs/adr/0026-expand-subscription-platform-metadata.md:50]: @meandavejustice Could you add an issue in the FXA-3655 epic for this?  Thanks!

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (vbudhram) [inline docs/adr/0026-expand-subscription-platform-metadata.md:29]: Another possible option is Redis (probably bad one at that)

Pros
- Already in tech stack and familiarity
- Can be updated without a deploy

Cons
- No UI for management
- Might need ops to manually update entires
- Would need to be populated/seeded with entries on startup

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (vbudhram) [inline docs/adr/0026-expand-subscription-platform-metadata.md:67]: We might need to do a migration to add these entries?

**Review by vbudhram — APPROVED** [review]
@chenba Makes sense to me 👍🏽

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (chenba) [reply docs/adr/0026-expand-subscription-platform-metadata.md:29]: Thanks, I'll include it in the final edit.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (chenba) [reply docs/adr/0026-expand-subscription-platform-metadata.md:67]: I was thinking that once we have a table, inserts and updates wouldn't require migrations.  Granted, we currently do not have a process of directly inserting and updating records on prod.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (biancadanforth) [inline docs/adr/0026-expand-subscription-platform-metadata.md:52]: What do you mean here? Can you give an example or how it may not be granular enough?

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (biancadanforth) [inline docs/adr/0026-expand-subscription-platform-metadata.md:54]: This is a risk people often overlook; I'm glad you mentioned it. I guess we also have this risk with Stripe metadata currently, but at least as noted above, we could validate the data before it is saved.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (biancadanforth) [inline docs/adr/0026-expand-subscription-platform-metadata.md:80]: I have heard there is a lot of historical evidence to suggest it's not an easy sell.

**Review by biancadanforth — COMMENTED** [review]
Thank you Barry for thinking about this. I added some non-blocking questions mostly along the line of seeking some more context from the perspective of someone who maybe isn't on our team.

Overall, I think this is the right technology choice -- especially concerning data validation before saving edits and not requiring deploys.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (bbangert) [reply docs/adr/0026-expand-subscription-platform-metadata.md:52]: The ACL capability in Firestore actually looks more advanced and granular than MySQL: https://firebase.google.com/docs/rules

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (bbangert) [reply docs/adr/0026-expand-subscription-platform-metadata.md:54]: We can still request updates go through our team, so we don't have to allow live data changes for PMs.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (chenba) [reply docs/adr/0026-expand-subscription-platform-metadata.md:13]: > 
> 
> > The Subscription Platform uses [Stripe Metadata][stripe-metadata] for numerous
> > purposes.
> 
> Can you list a few purposes here to be more explicit? Who updates this metadata today?

Sure.

> 
> > However, Stripe metadata has a limit on the number of values that we will reach in the near future.
> 
> Maybe we can link to the Stripe docs here that discuss that limit?

Yep.


> How close are we to this limit today? What is the timeline in which we expect to exceed that limit? More broadly, are there other non-idealities to relying on Stripe metadata than this limit (e.g. metadata validation)?
> 

The number of metadata keys used is product and pricing plan dependent.  I could give some kind of napkin sketch math here if you think that'd be helpful.

As for the timeline, that's purely a Product thing; I don't truly know.  Would that affect the decision?

Validation can be a whole separate topic altogether, regardless of storage, transport, access control, etc.  We can discuss more in a future jira (the one I requested Dave to file).  It's not a basis for the decision here is what I think I'm trying to say.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (chenba) [reply docs/adr/0026-expand-subscription-platform-metadata.md:52]: I could remove that bullet point if the Firestore ACL rules meet our needs.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (chenba) [reply docs/adr/0026-expand-subscription-platform-metadata.md:54]: Even if someone from our team handles the editing, there still no review and approval process.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (lmorchard) [reply docs/adr/0026-expand-subscription-platform-metadata.md:29]: Not that folks are necessarily looking for yet another option, but how about...

[Kinto](https://docs.kinto-storage.org/en/latest/)

Pros

* Already deployed and in use at Mozilla for Remote Settings in Firefox
* Data can be updated without FxA deployments
* Multiple collections per Kinto instance - e.g. dev / stage / prod product sets
* Record creation UI via JSON schema - maybe friendlier than raw JSON
* Implements a review process for changes with approve / reject flow
* Maintains a changelog of updates and who was involved
* Has ACLs and groups for read / write / review access

Cons

* Backed by postgres, while FxA uses mysql
* Not as turn-key to set up as something like Firestore
* Yet another service & a new dependency for FxA

Trying to think of other cons

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (lmorchard) [inline docs/adr/0026-expand-subscription-platform-metadata.md:25]: Meant to hang this comment over here - not that folks are necessarily looking for yet another option, but how about...

[Kinto](https://docs.kinto-storage.org/en/latest/)

Pros

* Already deployed and in use at Mozilla for Remote Settings in Firefox
* Data can be updated without FxA deployments
* Multiple collections per Kinto instance - e.g. dev / stage / prod product sets
* Record creation UI and data validation via JSON schema - maybe friendlier than raw JSON
* Implements a review process for pending changes with approve / reject flow
* Maintains a changelog of updates and who was involved
* Has ACLs and groups for read / write / review access

Cons

* Backed by postgres, while FxA uses mysql
* Not as turn-key to set up as something like Firestore
* Yet another service & a new dependency for FxA

Trying to think of other cons

**Review by lmorchard — COMMENTED** [review]
Just had a notion that [Kinto could be interesting](https://docs.kinto-storage.org/en/latest/overview.html#key-features) to consider here, after working with it for a bit over on Experimenter

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (chenba) [reply docs/adr/0026-expand-subscription-platform-metadata.md:25]: Thanks, Les!  With those cons, I can imagine jbuck being in favor of Firestore.  The pros sound great though.

I'll incorporate it into the final edit. 👍

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (bbangert) [reply docs/adr/0026-expand-subscription-platform-metadata.md:54]: If we wanted a full review/approval process, ops can manage Firestore documents. They are now doing this for the firestore documents backing event-broker. So there's a formalized process if we want one.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (bbangert) [reply docs/adr/0026-expand-subscription-platform-metadata.md:52]: I think if anything, the ACL capability in Firestore should be a Pro. It's fully programmatic as well, allowing functions to be used to gauge level of access on a per-field and per document basis for each action.

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (biancadanforth) [reply docs/adr/0026-expand-subscription-platform-metadata.md:13]: > The number of metadata keys used is product and pricing plan dependent. I could give some kind of napkin sketch math here if you think that'd be helpful.
>
> As for the timeline, that's purely a Product thing; I don't truly know. Would that affect the decision?

Hm... no I don't think that'd be helpful here. I'm just curious about these things from an urgency perspective, but you're right that this sort of information is not the most relevant to include here.

> Validation can be a whole separate topic altogether, regardless of storage, transport, access control, etc. We can discuss more in a future jira (the one I requested Dave to file). It's not a basis for the decision here is what I think I'm trying to say.

Why wouldn't validation (before saving) be a basis for a decision here -- particularly if we are planning to go with an approach that does not provide a direct review/approval process? That seems like a "pro" to me, even if it isn't part of an initial implementation... Is it because pretty much all the approaches we have thus far considered have some mechanism to validate the data (unlike the Stripe metadata approach we are using currently)?

Comment (chenba) [reply docs/adr/0026-expand-subscription-platform-metadata.md:13]: > Why wouldn't validation (before saving) be a basis for a decision here -- particularly if we are planning to go with an approach that does not provide a direct review/approval process? ...

The validation rules are from us, regardless the form they take, the way they are managed, or the context in which they are executed.  The feature that Firestore provide seems great (at least from some quick reading of their docs); I'll add it as a pro.  But generally, none of the options would prevent us from performing validation... or, what you said, and said it better.

**Review by dannycoates — APPROVED** [review]
SGTM

File: docs/adr/0026-expand-subscription-platform-metadata.md
Comment (chenba) [reply docs/adr/0026-expand-subscription-platform-metadata.md:13]: > 
> 
> > Maybe we can link to the Stripe docs here that discuss that limit?
> 
> Yep.
> 
I just realized that the doc linked in the ADR discusses that limit.  There isn't another page to link.

**Comment by chenba** [comment]
I pushed a final draft with everyone included in the deciders list and incorporated the review feedback.

Thank you all!

**Pull request merged**
